go/nbs/fd_cache.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2017 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package nbs
+
+import (
+	"os"
+	"sort"
+	"sync"
+
+	"github.com/attic-labs/noms/go/d"
+)
+
+func newFDCache(targetSize int) *fdCache {
+	return &fdCache{targetSize: targetSize, cache: map[string]fdCacheEntry{}}
+}
+
+// fdCache ref-counts open file descriptors, but doesn't keep a hard cap on
+// the number of open files. Once the cache's target size is exceeded, opening
+// a new file causes the cache to try to get the cache back to the target size
+// by closing fds with zero refs. If there aren't enough such fds, fdCache
+// gives up and tries again next time a caller refs a file.
+type fdCache struct {
+	targetSize int
+	mu         sync.Mutex
+	cache      map[string]fdCacheEntry
+}
+
+type fdCacheEntry struct {
+	refCount uint32
+	f        *os.File
+}
+
+// RefFile returns an opened *os.File for the file at |path|, or an error
+// indicating why the file could not be opened. If the cache already had an
+// entry for |path|, RefFile increments its refcount and returns the cached
+// pointer. If not, it opens the file and caches the pointer for others to
+// use.
+// If reffing a currently unopened file causes the cache to grow beyond
+// |fc.targetSize|, RefFile makes a best effort to shrink the
+// cache by dumping entries with a zero refcount. If there aren't enough zero
+// refcount entries to drop to get the cache back to |fc.targetSize|, the
+// cache will remain over |fc.targetSize| until the next call to RefFile().
+func (fc *fdCache) RefFile(path string) (f *os.File, err error) {
+	f, err = fc.checkCache(path)
+	if f != nil || err != nil {
+		return f, err
+	}
+
+	// Very much want this to be outside the lock
+	f, err = os.Open(path)
+
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+	defer fc.maybeShrinkCache()
+	if err != nil {
+		return nil, err
+	}
+	fc.cache[path] = fdCacheEntry{f: f, refCount: 1}
+	return f, nil
+}
+
+func (fc *fdCache) checkCache(path string) (f *os.File, err error) {
```

Comment:
check cache isn't really a great name for this because its incrementing ref count on a file. there's only one caller for this. what about just inlining this in `RefFile` and move GCing open fds if neccesary to Unref()?